### DISCUSSION TITLE
How do I use decorators in nuxt(ssr = false)?

### DISCUSSION THREAD

**Discussion by void-soul** [nuxt/nuxt#28137] [Questions]
How can I use `decorators` when `ssr=false` is set?
To make it easier to express, I created a new project without modifying anything except the following:
`nuxt.config.ts`
```
// https://nuxt.com/docs/api/configuration/nuxt-config
export default defineNuxtConfig({
  compatibilityDate: '2024-04-03',
  devtools: { enabled: true },
  ssr: false
})
```

`tsconfig.json`
```
{
  // https://nuxt.com/docs/guide/concepts/typescript
  "extends": "./.nuxt/tsconfig.json",
  "compilerOptions": {
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
  }
}
```
A simple decorator:

`utils/log.ts`
```
export function Log(target: any, key: string, descriptor: PropertyDescriptor) {
    const originalMethod = descriptor.value;

    descriptor.value = function (...args: any[]) {
        console.log(`Calling "${key}" with`, args);
        return originalMethod.apply(this, args);
    };

    return descriptor;
}
export class T {
    @Log
    test() {

    }
}
```

`pages/index.vue`

```
<template>
  <div>
    test
  </div>
</template>

<script lang="ts" setup>
new T().test();
</script>
```

After starting the project, accessing the home page prompts an error:
```
500
Invalid or unexpected token
```

**Comment by angelhdzmultimedia** [comment]
`nuxt.config.ts`:
```ts
...
vite: {
 esbuild: {
  tsconfigRaw: {
        compilerOptions: {
          experimentalDecorators: true,
          emitDecoratorMetadata: true,
        }
      }
 }
}
```

**Reply by ayu-sh-kr** [reply]
Hey @angelhdzdev I don't understand why enabling the `experimentalDecorators=true` in the `tsconfig.json` doesn't work. But doing the same via the `nuxt.config` works. Can you explain a little about the Vite configuration in the `nuxt.config`

**Reply by angelhdzmultimedia** [reply]
> Hey @angelhdzdev I don't understand why enabling the `experimentalDecorators=true` in the `tsconfig.json` doesn't work. But doing the same via the `nuxt.config` works. Can you explain a little about the Vite configuration in the `nuxt.config`

In a Nuxt 3 project, configuring tsconfig options within the vite object in nuxt.config.ts is necessary when the specific TypeScript configurations are intended for Vite's processing during the build and dev process, not the global TypeScript setup. Here's why:

1. Vite-Specific Configuration

Vite uses its own TypeScript handling internally, often relying on esbuild for fast transpilation.

Vite does not read the tsconfig.json directly during its build or dev processes. Instead, you must specify relevant TypeScript options (like paths or strict mode) in the vite object in nuxt.config.ts.


2. Paths Mapping

Nuxt automatically generates a tsconfig.json in .nuxt/tsconfig.json based on your Nuxt project's structure and configurations. This includes vite-specific aliases and paths that Vite uses during runtime.

By adding TypeScript configuration in the vite block, you ensure alignment between your custom configuration and what Vite needs for module resolution.


3. Scoped Configuration

Configuring TypeScript inside vite allows you to define behavior that only applies during Vite's compilation phase, without affecting your global tsconfig.json.

This is useful when Vite requires special tweaks (e.g., resolving aliases differently) that are not relevant to other TypeScript tools like tsc or your IDE.


4. Avoid Overriding Nuxt's Automatic tsconfig.json

Nuxt 3 automatically manages a tsconfig.json file in .nuxt/tsconfig.json. This generated file integrates your app's settings, auto-imported components, composables, and modules.

By configuring TypeScript options inside vite, you avoid conflicting with or overriding Nuxt's default behavior, keeping your global tsconfig.json clean and focused on project-wide TypeScript settings.
